pkg/webhook/instance_admission.go

```diff
@@ -335,20 +343,79 @@ func triggeredByParameterUpdate(params []kudov1beta1.Parameter, ov *kudov1beta1.
+func changedParameterDefinitions(old, new map[string]string, newOv, oldOv *kudov1beta1.OperatorVersion, isUpgrade bool) ([]kudov1beta1.Parameter, error) {
+	changed, removed := kudov1beta1.RichParameterDiff(old, new)
+	changedDefs, err := kudov1beta1.GetParamDefinitions(changed, newOv)
 	if err != nil {
 		return nil, err
 	}
 
+	if !isUpgrade {
+		for _, p := range changedDefs {
+			if p.IsImmutable() {
+				return nil, fmt.Errorf("parameter '%s' is immutable but was changed from '%v' to '%v'", p.Name, old[p.Name], new[p.Name])
+			}
+		}
+	} else {
+		if err := validateOVUpgrade(new, newOv, oldOv); err != nil {
+			return nil, err
+		}
+	}
+
 	// we ignore the error for missing OV parameter definitions for removed parameters. For once, this is a valid use-case when
 	// upgrading an Instance (new OV might remove parameters), but the user can also manually edit current OV and remove parameters.
 	// while discouraged, this is still possible since OV is not immutable.
-	rpd, _ := kudov1beta1.GetParamDefinitions(r, ov)
+	removedDefs, _ := kudov1beta1.GetParamDefinitions(removed, newOv)
+
+	return append(changedDefs, removedDefs...), nil
+}
+
+func validateOVUpgrade(new map[string]string, newOv, oldOv *kudov1beta1.OperatorVersion) error {
+	isEqualImmutable := func(p1, p2 kudov1beta1.Parameter) bool {
+		return p1.IsImmutable() == p2.IsImmutable()
+	}
+
+	for _, changedParam := range kudov1beta1.GetChangedParameters(oldOv, newOv, isEqualImmutable) {
+		if changedParam.IsImmutable() {
+			// Param was changed from Mutable to Immutable - We need to make sure we have a value set
+			if _, ok := new[changedParam.Name]; !ok {
+				return fmt.Errorf("parameter '%s' was changed to immutable in operator version %s but no value was provided", changedParam.Name, newOv.Name)
+			}
+		}
+		// else {
+		// Param was changed from Immutable to Mutable - Nothing to validate
+		// }
+	}
+
+	for _, removedParam := range kudov1beta1.GetRemovedParameters(oldOv, newOv) {
+		// Param was removed - Remove the value from the parameters if it exists
+		delete(new, removedParam.Name)
+	}
 
-	return append(cpd, rpd...), nil
+	for _, addedParam := range kudov1beta1.GetAddedParameters(oldOv, newOv) {
+		if addedParam.IsImmutable() {
+			// A new immutable param was added
+			if _, ok := new[addedParam.Name]; !ok {
+				if addedParam.HasDefault() {
+					return fmt.Errorf("parameter '%s' was added in operator version %s, but no value was provided (default would be %s)", addedParam.Name, newOv.Name, *addedParam.Default)
+				}
+				return fmt.Errorf("parameter '%s' was added in operator version %s, but no value was provided", addedParam.Name, newOv.Name)
+			}
+		}
+	}
+	return nil
+}
+
+// setImmutableParameterDefaults sets the default values for immutable parameters into the instances parameter map
+func setImmutableParameterDefaults(ov *kudov1beta1.OperatorVersion, instance *kudov1beta1.Instance) {
```

Comment:
I wonder if this is not better suited in the kudoctl 🤔  I was trying to minimize the mutations the IAW is doing to the instance (basically limiting it to `PlanExecution`). If we start to set the immutable defaults here, why not the _normal_ defaults too?

Reply:
I don't know, why don't we set the defaults here?

I think we should do as little as possible in `kudoctl` - I expect people to more and more use the CRDs directly, either with KUDO shim, or directly modifying the `instance` CRDs. From my perspective the `kudoctl` should only provide convenience functionality but nothing that directly affects the KUDO core funktionality.